pep-9999.rst

```diff
@@ -0,0 +1,868 @@
+------------------------------------
+
+Regardless whether support for freezing is opt-in or opt-out some
+types will not be freezable. (Example such types include IO types
+like file handles, and caches -- as opposed to the cached
+objects.) This raises the question how to handle failure to freeze
+an object graph. Consider the object graph ``o1 --> o2 --> o3``
+where ``o1`` and ``o3`` can be frozen, but ``o2`` cannot. What are
+the possible behaviours of ``freeze(o1)``?
+
+1. Freeze fails partially. All subgraphs which could be frozen
+   entirely remain frozen. Remaining objects remain mutable. In
+   our example, ``o3`` remains frozen but ``o1`` and ``o2`` remain
+   mutable. This preserves strict immutability. The exception
+   thrown by the failing ``freeze(o1)`` call will contain ``o2``
+   (the place that caused freezing to fail) and ``o1`` (the object
+   in the graph that holds on to the failing object) to facilitate
+   debugging.
+2. **Rejected alternative**: Freeze fails completely. In the strict
+   interpretation of deep immutability, freezing ``o1`` is not
+   possible because ``o1`` contains a reference to an un-freezable
+   object ``o2``. In this scenario, the object graph ``o1 --> o2
+   --> o3`` remains mutable and ``freeze(o1)`` raises an exception
+   when the object graph traversal encounters ``o2``.
+3. **Rejected alternative**: Freeze succeeds by altering the
+   graph. In this example removing ``o2`` from the graph or
+   swapping out ``o2`` for a placeholder object to be able to
+   freeze the graph. This alternative becomes complicated both to
+   reason about from a user's perspective, and to implement when
+   ``o2`` is referenced multiple times.
+4. **Rejected alternative**: Permit the user to choose between
+   alternatives 1) and 3) at use-site. In this case, the
+   ``freeze`` function takes an optional 2nd argument ``strict``
+   which must either be ``True`` or ``False``. In the first case,
+   ``freeze`` behaves as in alternative 1), in the second case,
+   it behaves as in alternative 2). We could further track whether
+   an object is strictly immutable or not in order to prevent
+   non-strictly immutable objects to participate in operations
+   which require strictness. This adds additional complexity to
+   the implementation, and also for the user.
+   
+This PEP proposes following alternative 1, where freezing either
+succeeds or fails completely.
+
+
+New Obligations on C Extensions
+-------------------------------
+
+Due to the opt-in decision, there are no *obligations* for C
+extensions that does not want to add support for immutability. 
+
+Because our implementation builds on information available to the CPython
+cycle detector, types defined through C code will support immutability 
+"out of the box" as long as they use Python standard types to store
+data and uses the built-in functions of these types to modify the data.
+
+To make its instances freezable, types using C extensions must
```

Comment:
Just as a note: this is only true if the C extension adds new functionality implemented in C (which isn't always the case). It is a subtle point but worth making. If you construct a C type using freezable metaclasses it will itself be freezable.